### PR TITLE
fix some shellcheck warnings

### PR DESCRIPTION
**Pull request by vnherdeiro**
Contributes to #6498

Following this attempt https://github.com/microsoft/LightGBM/pull/6548 at contributing to LightGBM and @jameslamb guidance I am now trying to help with fixing the shellcheck warnings for all the bash scripts inside the package (at least I think so). 

Screenshot to show that the warnings are gone in my local environment:
Screenshot 2024-08-24 at 21 05 13

Fixing https://www.shellcheck.net/wiki/SC1091 errors (source activate $SOME_CONDA_ENV) by pointing at at /dev/null is quite unsatisfactory but I did not manage to come up with something better.

Welcoming any feedback!

### CODE REVIEW

**Comment by vnherdeiro** [comment]
@microsoft-github-policy-service agree

File: build-python.sh
Comment (vnherdeiro) [inline build-python.sh:357]: This raises spellcheck info level `SC2086 (info): Double quote to prevent globbing and word splitting.` but adding the double quotes breaks the command (BUILD_ARGS is empty) I don't understand the root cause so I leave it as it. This is repeated twice in the file.

**Comment by vnherdeiro** [comment]
I had to revert some of the changes due to breaking commands. Now we have 3 `SC2086 (info): Double quote to prevent globbing and word splitting. ` in build-python.sh. Pipelines seem to be passing (except the mac one, let me know if it's a regression please).

File: docs/build-docs.sh
Comment (jameslamb) [inline docs/build-docs.sh:23]:
```suggestion
# shellcheck source=/dev/null
```

I can tell from running `shellcheck` locally that this is intended to address this `shellcheck` error:

```text
^------^ SC1091 (info): Not following: activate was not specified as input (see shellcheck -x).
```

Can you please explain how `shellcheck source=/dev/null` works (hopefully with a link to documentation)? I'm not familiar with that that does.

File: docs/build-docs.sh
Comment (jameslamb) [inline docs/build-docs.sh:14]:
```suggestion
/bin/bash "${HOME}/miniforge.sh" -b -p "${CONDA}"
```

File: docs/build-docs.sh
Comment (jameslamb) [inline docs/build-docs.sh:11]:
```suggestion
    -o "${HOME}/miniforge.sh" \
```

File: .ci/rerun-workflow.sh
Comment (jameslamb) [inline .ci/rerun-workflow.sh:34]: Was this intended to address a specific `shellcheck` error, or is it just something you're proposing for stylistic similarity? I don't mind this change (I generally support wrapping interpolation in `${}` like this), just trying to understand the change.

File: .ci/check-python-dists.sh
Comment (jameslamb) [inline .ci/check-python-dists.sh:23]:
```suggestion
    check-wheel-contents "$(echo "${DIST_DIR}/"*.whl)" || exit 1
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for your help, and for taking the time to contribute to LightGBM!

A few comments:

### 1. The tool is called `shellcheck`, not `spellcheck`

I've changed that in the description and title.

### 2. Please limit this to just a few files

I do really appreciate your help here, but think this PR is too large to review. As it says in https://github.com/microsoft/LightGBM/issues/6498

> **It is not necessary to fix all of the problems in a single pull request.**

There are some changes in this PR that are more controversial and require a bit more discussion, and I don't want all of this work to be blocked on those discussions... let's please reduce this to a smaller set of files and work together on fixing all the `shellcheck` errors in them. Once this PR is merged, hopefully you'll be able to help with the other `shellcheck` errors in a follow-up PR.

Pease remove the changes to the following files:

* `.ci/test.sh`
* `.ci/test-r-package.sh`
* `.ci/setup.sh`
* `build-python.sh`
* `build-cran-package.sh`

Since you made this PR from a branch also called `master` on your fork (please do not do that in the future, as I asked in https://github.com/microsoft/LightGBM/pull/6548#issuecomment-2257475958), the easiest way to do that is to just go to https://github.com/microsoft/LightGBM, copy the file content, paste it in one your local clone, commit the changes, and push.

I've left suggestions on the changes to the other files.

### 3. Please use linking here on GitHub to connect the discussions

I've added a reference to #6498 in the description. In the future, when you contribute here and the work is related to other discussions, please do link them.

File: docs/build-docs.sh
Comment (vnherdeiro) [reply docs/build-docs.sh:23]: I read the error description on shellcheck https://www.shellcheck.net/wiki/SC1091 and my understanding is that a `source` command will point to a script file to executed at that location. Shellcheck would like to parse that file too to proof read it, the `#shellcheck source=PATH_TO_SCRIPT` is supposed to point to the path of the file and allow a static tool to ingest the script. Unfortunately here the file does not exist until your run L17 creating that `docs-env` environment, hence the check cannot happen. One way for silencing the error is to point to `/dev/null` which can be assumed to be empty and close the check. A global exception to the `1091` rule could be added to a `.shellcheck` file but there may be places where the check could be and should be applied.

**Comment by vnherdeiro** [comment]
@jameslamb Thanks for the extensive review. I have noted all of your points. Apologies for pushing from master, I did not notice I had forgotten to branch out beforehand. I have reversed the files you have asked me too, let's keep the changes for another PR. It think I have adressed all of your points so far.

File: docs/build-docs.sh
Comment (jameslamb) [reply docs/build-docs.sh:23]:
```suggestion
# shellcheck disable=SC1091
```

Ah got it, ok thanks for that! Instead of doing this, let's please just skip this particular error on this particular line, as described in https://www.shellcheck.net/wiki/Ignore.

That makes it clear which specific check from `shellcheck` we're working around.

File: .ci/rerun-workflow.sh
Comment (jameslamb) [reply .ci/rerun-workflow.sh]: ok just checking. This is fine, let's leave it (I do prefer the style with `${}`) but please don't touch other lines that aren't directly related to the purpose of the PR.

**Review by jameslamb — APPROVED** [review]
Thanks very much, these changes look good to me.

It won't matter because we squash-merge PRs here, but just so you know... most of the commits you pushed here were not correctly tied to your GitHub account. You can see https://github.com/microsoft/LightGBM/pull/3607#issuecomment-735358654 and the things it links to to learn more about that. If you plan to continue working on projects here on GitHub, I recommend adjusting your `.gitconfig` to ensure your commit are tied to your GitHub account.

File: .ci/check-python-dists.sh
Comment (jameslamb) [reply .ci/check-python-dists.sh]: In the future, please click "add suggestion to batch" and then eventually "commit suggestions" on these inline suggestion comments. That way, reviewers won't have to manually verify that you've applied the suggested changes, and GitHub will automatically mark the conversation "resolved" and collapse the threads.

**Comment by vnherdeiro** [comment]
Thanks for the approval @jameslamb 

Yeah I had started pushing from a git config not tied to my github email. One of the reasons I squashed all these commits together. Shouldn't happen again.

File: .ci/lint-cpp.sh
Comment (jameslamb) [inline .ci/lint-cpp.sh:13]: What was the purpose of these changes that you pushed after I approved? It looks to me like they result in the files we care about all being ignored.

```text
Ignoring file: ./cmake/IntegratedOpenCL.cmake ./cmake/Sanitizer.cmake ./cmake/modules/FindLibR.cmake ./CMakeLists.txt 
```

([build link](https://github.com/microsoft/LightGBM/actions/runs/10638315268/job/29504133350?pr=6621))

If this was from that `shellcheck` warning about assigning the result of `find` to a variable... let's just use `find -exec` instead. Please try this:

```shell
find \
    . \
    -type f \
    \( -name CMakeLists.txt -o -name '*.cmake' \) \
    -not -path './external_libs/*' \
    -exec cmakelint \
        --linelength=120 \
        --filter=-convention/filename,-package/stdargs,-readability/wonkycase \
        {} \+
```

I tested this locally (on my mac) and found it worked well. This listed the expected files:

```shell
find \
    . \
    -type f \
    \( -name CMakeLists.txt -o -name '*.cmake' \) \
    -not -path './external_libs/*'
```

```text
./CMakeLists.txt
./cmake/Sanitizer.cmake
./cmake/IntegratedOpenCL.cmake
./cmake/modules/FindLibR.cmake
```

And the full statement did raise warnings about CMake formatting when I changed some code in those scripts.

**Review by jameslamb — CHANGES_REQUESTED** [review]
I re-reviewed the changes you pushed after my last approval and found an issue with them. Please see the suggestion I left here.

File: .ci/lint-cpp.sh
Comment (vnherdeiro) [reply .ci/lint-cpp.sh:13]: @jameslamb First I did not mean to creep in a change after your approval, I am used to MR/PR settings in which a code change requires a re-approval. I changed the line because the `cmakelint` execution was failing the CI. I traced it to the file path arguments being passed with '\n' joins instead of single-whitespace. I changed the `find + (negative) grep` command for a more efficient `find` with a ignore path logic.

I missed the log line saying that the files were ignored, and I don't understand why it happens. I did think of uxing `exec` but I guessed the command would get quite long, hence less readable, and some one may want to print the value of `$cmake_files` for debugging purposes. I will accept your change, thanks for proofing!

Comment (jameslamb) [reply .ci/lint-cpp.sh]: ohhhhhhh thank you so much for the thorough explanation!

There shouldn't be a `./build/` directory in CI (because we don't actually run a CMake build in the `lint` CI task), but it is likely that there could be one lying around when you run this locally if you've also been doing other development in the project.

Keeping `./cmake/*.cmake` for now is fine... this is how the linting here worked before (e.g. is not a change made by your PR) and all of the `*.cmake` files we currently want to lint do happen to be in that directory).

**Review by jameslamb — APPROVED** [review]
thanks!

**Pull request merged**
